dom/job/jobstate.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 The Drivers Cooperative. All rights reserved.
+// Use of this source code is governed by a dual
+// license that can be found in the LICENSE file.
+
+package job
+
+import (
+	"strings"
+
+	"github.com/openmarketplaceengine/openmarketplaceengine/dao"
+)
+
+type State int
+
+const (
+	StateUnspecified      State = 0
+	StateAvailable        State = 1
+	StateGoToPickup       State = 2
+	StateNearPickup       State = 3
+	StateAwaitingPickup   State = 4
+	StateOnTrip           State = 5
+	StateNearDestination  State = 6
+	StateAtDestination    State = 7
+	StateComplete         State = 8
+	StateCompleteEarly    State = 9
+	StateCanceled         State = 10
+	StateCanceledByDriver State = 11
+	StateCanceledByRider  State = 12
+	StateRiderNoShow      State = 13
+)
+
+var stateString = [...]string{
+	StateUnspecified:      "UNSPECIFIED",
+	StateAvailable:        "AVAILABLE",
+	StateGoToPickup:       "GO_TO_PICKUP",
+	StateNearPickup:       "NEAR_PICKUP",
+	StateAwaitingPickup:   "AWAITING_PICKUP",
+	StateOnTrip:           "ON_TRIP",
+	StateNearDestination:  "NEAR_DESTINATION",
+	StateAtDestination:    "AT_DESTINATION",
+	StateComplete:         "COMPLETE",
+	StateCompleteEarly:    "COMPLETE_EARLY",
+	StateCanceled:         "CANCELED",
+	StateCanceledByDriver: "CANCELED_BY_DRIVER",
+	StateCanceledByRider:  "CANCELED_BY_RIDER",
+	StateRiderNoShow:      "RIDER_NO_SHOW",
+}
+
+var stringState = buildStateMap()
+
+//-----------------------------------------------------------------------------
+
+func StateFromString(s string, prefix ...string) (state State, found bool) {
+	if len(prefix) > 0 {
+		p := strings.Join(prefix, "")
+		if len(p) > 0 && strings.HasPrefix(s, p) {
+			s = s[len(p):]
+		}
+	}
+	s = strings.ToUpper(s)
+	state, found = stringState[s]
+	return
+}
+
+//-----------------------------------------------------------------------------
+
+func StateFromNumber(n int) (state State, found bool) {
+	if n >= 0 && n < len(stateString) {
+		state = State(n)
+		found = true
+	}
+	return
+}
```

Comment:
`jobstate.pb.go` contains maps we can use for this:
* `JobState_name` (`map[int32]string`)
* `JobState_value` (`map[string]int32`)

Reply:
I don't want to tide dom package to the gRPC api package and the names have prefix anyways.

Reply:
I think separating the `dom` package from any auto-generated proto code makes sense.

This field is ultimately either going to be a string or an integer. (I think we should go with integer as the column since it's less space and we don't have to worry about the `JOB_STATE_` prefix).

If we do choose to model it with a string column, any kind of translation from human-readable string (e.g., `"JOB_STATE_NEAR_DESTINATION"` to its less verbose counterpart `"NEAR_DESTINATION"`) could happen higher-up in the call stack, in the controller/handler layer.

Reply:
I thought about string vs integer for state storage. Yes, number is more efficient, but string is human readable and easier for custom queries in the future.

Bronx uses string for state storage.

If we all agree to migrate to the int storage, I'll open a new PR for this.

Reply:
I think we should encourage humans to query OME via a UI or the gRPC layer rather than via direct access to the database.

In Bronx I think we went with strings because there's lots of complexity and domains, and people are using Hasura console or psql and need something human-readable. We didn't have anything like a protobuf enum, which is a stable, unbreakable mapping from strings to numbers.

The scope of OME feels much more constrained than Bronx. At some point in the not too distant future we could start building a simple, bespoke UI for administration (searching, filtering, mutating, etc) — akin to other open-source tools (e.g., k8s, Elasticsearch, Jaeger, CNCF projects, etc). There would be human-readability in the UI and in the gRPC layer, just not in the db layer.

Reply:
> I think we should encourage humans to query OME via a UI or the gRPC layer rather than via direct access to the database.

Ideally, yes. Primarily direct access is required for debugging. Still not sure we should go with int for state storage. Anyways, if we decide so, it will be a separate PR.